src/linear_algebra/tensor_algebra.lean

```diff
@@ -0,0 +1,350 @@
+/-
+Copyright (c) 2020 Adam Topaz. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Author: Adam Topaz.
+-/
+
+import ring_theory.algebra
+import linear_algebra
+
+/-!
+# Tensor Algebras
+
+Given a commutative ring R, and an R-module M, we construct the tensor algebra of M.
+This is the free R-algebra generated by the module M.
+
+## Notation
+
+1. `tensor_algebra R M` is the tensor algebra itself.
+  It is endowed with an R-algebra structure.
+2. `univ R M` is the R-linear map `M → tensor_algebra R M`.
+3. Given a linear map `f : M → A` to an R-algebra A,
+  `lift R M f` is the lift of `f` to an R-algebra morphism
+  `tensor_algebra R M → A`.
+
+## Theorems
+
+1. `univ_comp_lift` states that the composition
+  `(lift R M f) ∘ (univ R M)` is identical to `f`.
+2. `lift_unique` states that whenever an R-algebra morphism
+  `g : tensor_algebra R M → A` is given whose composision with `univ R M` is `f`,
+  then one has `g = lift R M f`.
+-/
+
+universes u1 u2 u3
+
+variables (R : Type u1) [comm_ring R]
+variables (M : Type u2) [add_comm_group M] [module R M]
+
+namespace tensor_algebra
+
+-- This inductive type is used to express representatives
+-- of the tensor algebra.
+@[nolint doc_blame]
+inductive pre
+| of : M → pre
+| of_scalar : R → pre
+| add : pre → pre → pre
+| mul : pre → pre → pre
+
+namespace pre
+
+instance : inhabited (pre R M) := ⟨of_scalar 0⟩
+
+-- These instances are only used to simplify the notation.
+@[nolint doc_blame]
+def has_coe_module : has_coe M (pre R M) := ⟨of⟩
+@[nolint doc_blame]
+def has_coe_ring : has_coe R (pre R M) := ⟨of_scalar⟩
+@[nolint doc_blame]
+def has_mul : has_mul (pre R M) := ⟨mul⟩
+@[nolint doc_blame]
+def has_add : has_add (pre R M) := ⟨add⟩
+@[nolint doc_blame]
+def has_zero : has_zero (pre R M) := ⟨of_scalar 0⟩
+@[nolint doc_blame]
+def has_one : has_one (pre R M) := ⟨of_scalar 1⟩
+@[nolint doc_blame]
+def has_scalar : has_scalar R (pre R M) := ⟨λ r m, mul (of_scalar r) m⟩
+@[nolint doc_blame]
+def has_neg : has_neg (pre R M) := ⟨mul (of_scalar (-1))⟩
+
+end pre
+
+local attribute [instance]
+  pre.has_coe_module pre.has_coe_ring pre.has_mul pre.has_add pre.has_zero
+  pre.has_one pre.has_scalar pre.has_neg
+
+@[nolint doc_blame]
+def lift_fun {A : Type u3} [ring A] [algebra R A] (f : M →ₗ[R] A) : pre R M → A :=
```

Comment:
I think it would be good to write short docstrings for these definitions. If anyone needs to refactor this code 6 months (or 3 years) from now, they will be very happy.

Reply:
Sure, I can do that. I was considering `lift_fun` as just an auxiliary thing needed to define `lift`, but it's no problem to add a docstring for this as well.